### PR TITLE
 YAML implicit resolver of ranges: #687

### PR DESCRIPTION
**Pull request by chinmay0301**
This PR is in reference to the issue #687 
Identified the following set of features from the test example which we can resolve implicitly: 
1) `spectrum` 
2) `supernova` 

Allowing input in the YAML file in the format : `black_body_sampling: 1 1000000 1.e+6 angstrom angstrom` 
corresponding to the labels `start`, `stop`, `num`, and the units of start and stop correspondingly. 

Ensured that even in the inputs are given in the dictionary format as they are currently being given, the code still works. Can implement comma separated inputs also if required. Currently changed only the `widgets.py` file. Please let me know if changes are to be reflected in some other file also

### CODE REVIEW

**Comment by chinmay0301** [comment]
@ftsamis  can you have a look at it once?

**Comment by yeganer** [comment]
It looks like these changes won't implement the desired functionality.
Have you tried running the tardis_example.yml with the new syntax?

**Comment by ftsamis** [comment]
I'd say @yeganer is correct. yaml implicit resolvers are a very specific thing and unrelated with the GUI code. Make a search and see how they work.

**Comment by chinmay0301** [comment]
But what we essentially want is to decompose a string input into the corresponding key values for the dict element `black_body_sampling`right? So why is it wrong to just decompose it using a python code in gui where we are comparing the predefined yaml file with the input yaml file?

**Comment by yeganer** [comment]
Our config validation isn't done by the gui but by `tardis.io.config_reader`. The gui doesn't work with the latest version anyway.

Instead of getting a dictionary with `'start', 'stop' and 'num'` we want to get a `np.linspace` when loading the config with yaml.

I suggest to test the changes with `spectrum` instead of `blackbody_sampling` as that's currently used.

**Comment by chinmay0301** [comment]
@yeganer how should I store the np array obtained? as in, should it be stored as `object['spectrum'] = np.linspace(start, stop, num)` ?

**Comment by chinmay0301** [comment]
I have'nt implemented the resolver yet, but essentially this is what we want to achieve right? That we can directly pass a `np.linspace` array to the `base.py` file in the montecarlo folder.

File: scripts/tardis [hunk 1]
```diff
@@ -6,15 +6,11 @@ import numpy as np
 
 import logging
 import argparse
-
-
-
```
Comment (ftsamis) [inline scripts/tardis]: Don't make changes to files or code that are out of scope for your pull request.

File: scripts/tardis [hunk 2]
```diff
@@ -6,15 +6,11 @@ import numpy as np
...
 tardis_description =\
 """
 TARDIS Supernova Montecarlo Radiative transfer code
 
 """
-
```
Comment (ftsamis) [inline scripts/tardis]: Don't make changes to files or code that are out of scope for your pull request.

File: tardis/gui/widgets.py [hunk 3]
```diff
@@ -256,9 +256,9 @@ def __init__(self, yamlconfigfile, parent=None):
                           'no_of_packets':[True, None],
                           'iterations':[True, None],
                           'black_body_sampling':{
-                                                    'start': '1 angstrom',
-                                                    'stop': '1000000 angstrom',
-                                                    'num': '1.e+6',
+                                                   'start': '1 angstrom',
+                                                   'stop': '1000000 angstrom',
+                                                   'num': '1.e+6',
```
Comment (ftsamis) [inline tardis/gui/widgets.py]: Don't make changes to files or code that are out of scope for your pull request.

File: tardis/io/config_reader.py [hunk 4]
```diff
@@ -6,9 +6,9 @@
 
 import tardis
 from tardis.io import config_validator
-from tardis.io.util import YAMLLoader, yaml_load_file
+from tardis.io.util import YAMLLoader, yaml_load_file, quantity_from_str
+from tardis.util import quantity_linspace
 
-pp = pprint.PrettyPrinter(indent=4)
```
Comment (ftsamis) [inline tardis/io/config_reader.py]: Don't make changes to files or code that are out of scope for your pull request.

File: tardis/io/util.py [hunk 5]
```diff
@@ -65,7 +66,18 @@ class YAMLLoader(yaml.Loader):
     A custom YAML loader containing all the constructors required
     to properly parse the tardis configuration.
     """
-
+    def demo_constructor(self, node):
+    	data = self.construct_scalar(node)
+    	return data
```
Comment (ftsamis) [inline tardis/io/util.py]: What is `demo_constructor`? It doesn't seem to serve any purpose to me.

File: tardis/io/util.py [hunk 6]
```diff
@@ -65,7 +66,18 @@ class YAMLLoader(yaml.Loader):
...
+    def spectrum_constructor(self, node):
+    	"""
+    	A constructor to convert spectrum values to nparray
+    	"""
+    	data = self.construct_scalar(node)
+    	data = data.split('%')
+        start = u.Quantity(data[0], data[3])
+        stop = u.Quantity(data[1], data[3])
+        return data
```
Comment (ftsamis) [inline tardis/io/util.py]: - Indentation seems wrong
- `spectrum_constructor` is way too restrictive for a name that constructs ranges in general.
- You create start and stop local variables without using them again.
- Your delimiting scheme seems different from that specified in #687.
- You seem to ignore the `num` property, you don't handle cases where the unit is omitted/not required.
- You return a list and not a linspace.
- The `spectrum_constructor` doesn't seem to be used at all in the end.

File: tardis/io/util.py [hunk 7]
```diff
@@ -90,13 +102,14 @@ def mapping_constructor(self, node):
         return OrderedDict(self.construct_pairs(node))
 
 YAMLLoader.add_constructor(u'!quantity', YAMLLoader.construct_quantity)
+YAMLLoader.add_constructor(u'!demo', YAMLLoader.demo_constructor)
```
Comment (ftsamis) [inline tardis/io/util.py]: We do not have or need any `!demo` properties.

File: tardis/io/util.py [hunk 8]
```diff
@@ -90,13 +102,14 @@ def mapping_constructor(self, node):
...
 YAMLLoader.add_implicit_resolver(u'!quantity',
                                  MockRegexPattern(quantity_from_str), None)
 YAMLLoader.add_implicit_resolver(u'tag:yaml.org,2002:float',
-                                 MockRegexPattern(float), None)
+                                                                          MockRegexPattern(float), None)
```
Comment (ftsamis) [inline tardis/io/util.py]: Don't make changes to files or code that are out of scope for your pull request.

File: tardis/io/util.py [hunk 9]
```diff
@@ -90,13 +102,14 @@ def mapping_constructor(self, node):
...
 YAMLLoader.add_constructor(yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
                            YAMLLoader.mapping_constructor)
-
+YAMLLoader.add_implicit_resolver(u'!demo', re.compile(r'^\d+%\d+%\d+%\d+$'))
```
Comment (ftsamis) [inline tardis/io/util.py]: Your `demo` constructor, except from an inappropriate name, doesn't seem to do anything useful.

File: tardis/montecarlo/base.py [hunk 10]
```diff
@@ -359,16 +359,16 @@ def from_config(cls, config):
...
-            config.spectrum.start.to('Hz', u.spectral()),
-            num=config.spectrum.num + 1)
-
-        return cls(seed=config.montecarlo.seed,
+        # spectrum_frequency = quantity_linspace(
+        #     config.spectrum.stop.to('Hz', u.spectral()),
+        #     config.spectrum.start.to('Hz', u.spectral()),
+        #     num=config.spectrum.num + 1)
+        spectrum_frequency = config.spectrum.array       
+	return cls(seed=config.montecarlo.seed,
```
Comment (ftsamis) [inline tardis/montecarlo/base.py]: - Don't comment code that you think is not needed, instead just remove it. 
- config.spectrum.array does not occur from any yaml constructor.
- The purpose was to have a linspace created directly from the `YAMLLoader` through e.g. `config.spectrum` but not only for spectrum.

File: tardis/montecarlo/base.py [hunk 11]
```diff
@@ -359,16 +359,16 @@ def from_config(cls, config):
...
                    spectrum_frequency=spectrum_frequency,
                    virtual_spectrum_range=config.montecarlo.virtual_spectrum_range,
                    sigma_thomson=sigma_thomson,
                    enable_reflective_inner_boundary=config.montecarlo.enable_reflective_inner_boundary,
                    inner_boundary_albedo=config.montecarlo.inner_boundary_albedo,
                    line_interaction_type=config.plasma.line_interaction_type,
-                   distance=config.supernova.get('distance', None))
\ No newline at end of file
+                   distance=config.supernova.get('distance', None))
```
Comment (ftsamis) [inline tardis/montecarlo/base.py]: Don't make changes to files or code that are out of scope for your pull request

File: tardis/util.py [hunk 12]
```diff
@@ -428,4 +428,4 @@ def quantity_linspace(start, stop, num, **kwargs):
         raise ValueError('Both start and stop need to be quantities with a '
                          'unit attribute')
 
-    return np.linspace(start.value, stop.to(start.unit).value, num, **kwargs) * start.unit
\ No newline at end of file
+    return np.linspace(start.value, stop.to(start.unit).value, num, **kwargs) * start.unit
```
Comment (ftsamis) [inline tardis/util.py]: Don't make changes to files or code that are out of scope for your pull request

File: tardis/io/config_reader.py [hunk 13]
```diff
@@ -217,6 +217,10 @@ def from_yaml(cls, fname, *args, **kwargs):
         try:
             yaml_dict = yaml_load_file(fname,
                                        loader=kwargs.pop('loader', YAMLLoader))
+            start = yaml_dict['spectrum']['start'] 
+ 			stop = yaml_dict['spectrum']['stop'] 
+ 			num = yaml_dict['spectrum']['num'] 
+ 			yaml_dict['spectrum']['array'] = quantity_linspace(start, stop, num+1)
```
Comment (ftsamis) [inline tardis/io/config_reader.py]: - Identation seems wrong.
- The bug report you say you are fixing wants the `start`, `stop`, and `num` properties removed, you still use them here and add another one.

File: tardis/util.py (hunk 12 above)
Comment (chinmay0301) [inline tardis/util.py]: I will look into it carefully this time. Sorry for so many errors.
